### PR TITLE
Test case for patch, plus fix to not swallow exceptions

### LINKED ISSUE #13652: Read CSV using c engine silently swallows useful exceptions
#### Code Sample, a copy-pastable example if possible

```python
import pandas
import codecs
import traceback
import sys

pandas.show_versions()

print("Pandas version: {}\n".format(pandas.__version__))
print("Python version: {}\n".format(sys.version_info))

def build_stream():
    filename = "tests/data/sauron.SHIFT_JIS.csv"
    handle = open(filename, "rb")
    codec = codecs.lookup("utf-8")
    utf8 = codecs.lookup('utf-8')
    # stream must be binary UTF8
    stream = codecs.StreamRecoder(
        handle, utf8.encode, utf8.decode, codec.streamreader, codec.streamwriter)
    return stream

def test_pandas(use_python):
    stream = build_stream()

    if use_python:
        engine = 'python'
    else:
        engine = 'c'
    df = pandas.read_csv(stream, engine=engine)


print("Showing stream error on read\n")
try:
    stream = build_stream()
    data = stream.read()
except Exception as exc:
    traceback.print_exc(file=sys.stdout)


print("Showing stream error on read_csv (python engine)\n")
try:
    stream = test_pandas(True)
except Exception as exc:
    traceback.print_exc(file=sys.stdout)


print("Showing missing stream error on read_csv (python c)\n")
try:
    stream = test_pandas(False)
except Exception as exc:
    traceback.print_exc(file=sys.stdout)
...
```
#### Actual Output

```

INSTALLED VERSIONS
------------------
commit: None
python: 3.4.3.final.0
python-bits: 64
OS: Linux
OS-release: 3.13.0-55-generic
machine: x86_64
processor: x86_64
byteorder: little
LC_ALL: None
LANG: en_US.UTF-8

pandas: 0.16.1
nose: 1.3.7
Cython: 0.24.0a0
numpy: 1.9.2
scipy: 0.16.0
statsmodels: None
IPython: 4.1.2
sphinx: None
patsy: None
dateutil: 2.4.2
pytz: 2015.4
bottleneck: None
tables: None
numexpr: 2.4.3
matplotlib: None
openpyxl: None
xlrd: 0.9.4
xlwt: None
xlsxwriter: None
lxml: None
bs4: None
html5lib: 0.9999999
httplib2: None
apiclient: None
sqlalchemy: None
pymysql: None
psycopg2: None
Pandas version: 0.16.1

Python version: sys.version_info(major=3, minor=4, micro=3, releaselevel='final', serial=0)

Showing stream error on read

Traceback (most recent call last):
  File "pandas_bug.py", line 34, in <module>
    data = stream.read()
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 798, in read
    data = self.reader.read(size)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 497, in read
    newchars, decodedbytes = self.decode(data, self.errors)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x83 in position 12: invalid start byte
Showing stream error on read_csv (python engine)

Traceback (most recent call last):
  File "pandas_bug.py", line 41, in <module>
    stream = test_pandas(True)
  File "pandas_bug.py", line 28, in test_pandas
    df = pandas.read_csv(stream, engine=engine)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 474, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 250, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 566, in __init__
    self._make_engine(self.engine)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 711, in _make_engine
    self._engine = klass(self.f, **self.options)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 1427, in __init__
    self.columns, self.num_original_columns = self._infer_columns()
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 1642, in _infer_columns
    line = self._buffered_line()
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 1769, in _buffered_line
    return self._next_line()
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 1800, in _next_line
    orig_line = next(self.data)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 820, in __next__
    data = next(self.reader)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 638, in __next__
    line = self.readline()
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 551, in readline
    data = self.read(readsize, firstline=True)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/codecs.py", line 497, in read
    newchars, decodedbytes = self.decode(data, self.errors)
UnicodeDecodeError: 'utf-8' codec can't decode byte 0x83 in position 0: invalid start byte
Showing missing stream error on read_csv (python c)

Traceback (most recent call last):
  File "pandas_bug.py", line 48, in <module>
    stream = test_pandas(False)
  File "pandas_bug.py", line 28, in test_pandas
    df = pandas.read_csv(stream, engine=engine)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 474, in parser_f
    return _read(filepath_or_buffer, kwds)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 250, in _read
    parser = TextFileReader(filepath_or_buffer, **kwds)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 566, in __init__
    self._make_engine(self.engine)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 705, in _make_engine
    self._engine = CParserWrapper(self.f, **self.options)
  File "/home/brett/.virtualenvs/datasets-service/lib/python3.4/site-packages/pandas/io/parsers.py", line 1072, in __init__
    self._reader = _parser.TextReader(src, **kwds)
  File "pandas/parser.pyx", line 509, in pandas.parser.TextReader.__cinit__ (pandas/parser.c:4732)
  File "pandas/parser.pyx", line 635, in pandas.parser.TextReader._get_header (pandas/parser.c:6244)
  File "pandas/parser.pyx", line 831, in pandas.parser.TextReader._tokenize_rows (pandas/parser.c:8275)
  File "pandas/parser.pyx", line 1742, in pandas.parser.raise_parser_error (pandas/parser.c:20691)
pandas.parser.CParserError: Error tokenizing data. C error: Calling read(nbytes) on source failed. Try engine='python'.
```
#### Expected Behavior

The C engine should behave like python engine. This should be possible by using PyErr_Occurred .

### PR DESCRIPTION
**Pull request by bdrosen96**
- [x] closes #13652
- [x] tests added / passed
- [x] passes `git diff upstream/master | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by gfyoung** [comment]
@bdrosen96 : l wonder if you might need another test to further illustrate that these (different) C errors are being generated.  The impression I would get from this test alone is that you were testing for an encoding error.  @jreback what do you think?

**Comment by bdrosen96** [comment]
With this fix, we get the same exception for both engines, where previously we got one exception for python engine and a different one for c engine (the cparser exception with no indication of the root cause of the failure)

I'm not sure what another deterministic stream error we could test here, unless we try to create a fake stream that errors after some number of bytes read.

**Comment by jreback** [comment]
fyi if you use the ``#`` on an issue it references automatically

**Comment by gfyoung** [comment]
@bdrosen : what I meant were there any other errors besides the encoding one that the C engine was previously swallowing that are now surfaced?

Just to emphasize: not a blocker, but a nice to have (i.e. a test for another error that was being swallowed)

File: pandas/parser.pyx [hunk 1]
```diff
@@ -1878,6 +1880,17 @@ cdef kh_float64_t* kset_float64_from_list(values) except NULL:
 
 
 cdef raise_parser_error(object base, parser_t *parser):
+    cdef:
+        object old_exc
+        PyObject *type, *value, *traceback
+    if PyErr_Occurred():
```
Comment (jreback) [inline pandas/parser.pyx]: https://docs.python.org/2/c-api/exceptions.html

do we need to decref this?

File: pandas/io/tests/parser/common.py [hunk 2]
```diff
@@ -45,6 +46,21 @@ def test_empty_decimal_marker(self):
         with tm.assertRaisesRegexp(ValueError, msg):
             self.read_csv(StringIO(data), decimal='')
 
+    def test_bad_stream_exception(self):
+        handle = open(self.csv_shiftjs, "rb")
```
Comment (jreback) [inline pandas/io/tests/parser/common.py]: add the issue number as a comment (and a 1-liner explaining what this test is catching)

File: pandas/io/tests/parser/common.py [hunk 3]
```diff
@@ -45,6 +46,21 @@ def test_empty_decimal_marker(self):
...
+        utf8 = codecs.lookup('utf-8')
+        # stream must be binary UTF8
+        stream = codecs.StreamRecoder(
+            handle, utf8.encode, utf8.decode, codec.streamreader,
+            codec.streamwriter)
+        if compat.PY3:
+            msg = "'utf-8' codec can't decode byte"
+        else:
+            msg = "'utf8' codec can't decode byte"
+        with tm.assertRaisesRegexp(UnicodeDecodeError, msg):
```
Comment (jreback) [inline pandas/io/tests/parser/common.py]: do we need to close the codec itself?

File: pandas/parser.pyx (hunk 1 above)
Comment (bdrosen96) [inline pandas/parser.pyx]: The return from PyErr_Occurred does not need to be decrefed. The return from PyErr_Fetch does, and is.

**Comment by bdrosen96** [comment]
Yes, there are other cases that would now be surfaced, but they would be hard to create a test for. For example:

1 HTTP request to get remote file as file like object (stream) passed to read-csv. Something goes wrong with network and exception is raised during read. We should now get a non generic error indicating network trouble rather than the generic read() error from before.

2 Same thing with FTP, HDFS, gluster, S3 or any other network type file operation.

3 Stream handle does decompression/unarchive (ie .csv.bz2 or .csv.gz, .csv.tar.gz, etc)

There are probably other cases that I am not aware of now.

File: pandas/io/tests/parser/data/sauron.SHIFT_JIS.csv [hunk 4]
```diff
@@ -0,0 +1,14 @@
+num, text
+1,�T�E�����iSauron�A�A�C�k�A�̑n���̎� - ��O�I3019�N3��25��j�́AJ�ER�ER�E�g�[���L���̒����𕑑�Ƃ�������w�z�r�b�g�̖`���x�w�w�֕���x�w�V���}�����̕���x�̓o��l���B
```
Comment (sinhrks) [inline pandas/io/tests/parser/data/sauron.SHIFT_JIS.csv]: not for the fix itself, but isn't the copylight still alive?

Comment (sinhrks) [inline pandas/io/tests/parser/data/sauron.SHIFT_JIS.csv]: ah maybe from wikipedia?

**Comment by jreback** [comment]
@bdrosen96 actually could easily test a network op on a non-existant file (and even an s3 op for the same, just give an invalid bucket).

**Comment by bdrosen96** [comment]
I don't think testing a network op on invalid bucket or non existant file would work since the exception would likely occur during open rather than read, so that would not help.

**Comment by jreback** [comment]
pls also add a whatsnew entry, you can put in API section as now better exceptions are passed thru the parser (plus I *think* the exception type could have changed), e.g. I think previously you would have gotten a ``CParserError`` (which is a ``ValueError``), now you can get other things. If you can posit a useful message would be great.

File: pandas/io/tests/parser/data/sauron.SHIFT_JIS.csv (hunk 4 above)
Comment (bdrosen96) [inline pandas/io/tests/parser/data/sauron.SHIFT_JIS.csv]: yes, this is annotations on the book, not the text of the book.

**Comment by bdrosen96** [comment]
Added whatsnew, closed codec recoder stream and added comment for test

**Comment by jreback** [comment]
thanks @bdrosen96

**Pull request closed without merging**